### PR TITLE
Start requiring PartitionEntry, pt1

### PR DESCRIPTION
**Pull request by esmadau**
## Description
Today, we sometimes use partition name and sometimes key. We often have partition in our code available, but the value is not always set.
This is a refactor that gets us closer to being able to rely on Partition being set and accessing its values always produces a value.
With IDP work continuing and supporting Update operation next, this work sets us up to have partition names that we can rely on to then use for blob paths in implementing that. It gets us closer to being able to use partition in other features as well without a heavier uplift per feature.
Also, we want to stop using context accessor below the api level so a propagation of Partition will also get us closer to this.
This came up as a request for cleanup while working on 104372. It is also an issue that impacted us where partition did not have values and so we saw NPEs ( see ticket # 105059).

The big changes are:
- Renaming PartitionEntry to Partition and inner fields, but maintaining PartitionEntry and its CreatedTime where needed for swagger and backwards compatibility
- Beginning of enforcing through use of ctors that partition name and key are always available when partition used in other classes such as InstanceIdentifier

## Related issues
Addresses [[AB#105251](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/105251)].

## Testing
Updated all tests to use Partition

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Core/Features/Model/VersionedInstanceIdentifier.cs [hunk 1]
```diff
@@ -27,7 +27,7 @@ public class VersionedInstanceIdentifier : InstanceIdentifier
         long version,
         int partitionKey = default,
         string partitionName = default)
-        : base(studyInstanceUid, seriesInstanceUid, sopInstanceUid, partitionKey, partitionName)
+        : base(studyInstanceUid, seriesInstanceUid, sopInstanceUid, new PartitionEntry(partitionKey, partitionName))
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Model/VersionedInstanceIdentifier.cs:30]: Should this ctors be symmetrical to the `InstanceIdentifier`?

File: src/Microsoft.Health.Dicom.Core/Features/Model/VersionedInstanceIdentifier.cs [hunk 2]
```diff
@@ -25,9 +25,9 @@ public class VersionedInstanceIdentifier : InstanceIdentifier
         string seriesInstanceUid,
         string sopInstanceUid,
         long version,
-        int partitionKey = default,
-        string partitionName = default)
-        : base(studyInstanceUid, seriesInstanceUid, sopInstanceUid, partitionKey, partitionName)
+        int partitionKey = Partition.DefaultKey,
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Model/VersionedInstanceIdentifier.cs:28]: Will we delete this ctor?

File: src/Microsoft.Health.Dicom.Core/Features/Partitioning/IPartitionStore.cs [hunk 3]
```diff
@@ -7,13 +7,13 @@
 using System.Threading;
 using System.Threading.Tasks;
 
-namespace Microsoft.Health.Dicom.Core.Features.Partition;
+namespace Microsoft.Health.Dicom.Core.Features.Partitioning;
 
 public interface IPartitionStore
 {
-    Task<PartitionEntry> AddPartitionAsync(string partitionName, CancellationToken cancellationToken = default);
+    Task<Partitioning.Partition> AddPartitionAsync(string partitionName, CancellationToken cancellationToken = default);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Partitioning/IPartitionStore.cs:14]: Does this conflict with something?

File: src/Microsoft.Health.Dicom.Core/Features/Partitioning/Partition.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+    public int Key { get; }
+
+    public string Name { get; }
+
+    public static Partition Default => new(DefaultKey, DefaultName);
+
+    public Partition(int key, string name)
+    {
+        Key = key;
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Partitioning/Partition.cs:24]: `> 0`?

File: src/Microsoft.Health.Dicom.Core/Features/Partitioning/Partition.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+{
+    public const string DefaultName = "Microsoft.Default";
+
+    public const int DefaultKey = 1;
+
+    public int Key { get; }
+
+    public string Name { get; }
+
+    public static Partition Default => new(DefaultKey, DefaultName);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Partitioning/Partition.cs:20]: Singleton:

```suggestion
    public static Partition Default { get; } = new(DefaultKey, DefaultName);
```

File: src/Microsoft.Health.Dicom.Core/Features/Partitioning/PartitionCache.cs [hunk 6]
```diff
@@ -8,9 +8,9 @@
 using Microsoft.Health.Dicom.Core.Configs;
 using Microsoft.Health.Dicom.Core.Features.Common;
 
-namespace Microsoft.Health.Dicom.Core.Features.Partition;
+namespace Microsoft.Health.Dicom.Core.Features.Partitioning;
 
-public class PartitionCache : EphemeralMemoryCache<string, PartitionEntry>
+public class PartitionCache : EphemeralMemoryCache<string, Partitioning.Partition>
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Partitioning/PartitionCache.cs:13]: Is there a type collision?

File: src/Microsoft.Health.Dicom.Core/Features/Store/IStoreResponseBuilder.cs [hunk 7]
```diff
@@ -28,8 +29,10 @@ public interface IStoreResponseBuilder
     /// <param name="storeValidationResult">Store validation errors and warnings</param>
     /// <param name="warningReasonCode">The warning reason code.</param>
     /// <param name="buildWarningSequence">Whether to build response warning sequence or not.</param>
+    /// <param name="partition">Data Partition entry</param>
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Store/IStoreResponseBuilder.cs:32]: nit: this is in the wrong order

File: src/Microsoft.Health.Dicom.Core/Features/Store/StoreResponseBuilder.cs [hunk 8]
```diff
@@ -97,7 +99,7 @@ private bool HasWarningReasonCode()
             _dataset.Add(referencedSopSequence);
         }
 
-        var dicomInstance = dicomDataset.ToInstanceIdentifier();
+        var dicomInstance = dicomDataset.ToInstanceIdentifier(partition);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Store/StoreResponseBuilder.cs:102]: This is good, and it further highlights some of the inconsistencies where the identifier is scoped to a partition, but the method below `ResolveRetrieveInstanceUri` fetches the partition from the request context.

Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Features/Store/StoreResponseBuilder.cs:102]: It may be worth including a `Debug.Assert` in that method to assert the partitions match

File: src/Microsoft.Health.Dicom.Core/Messages/Partitioning/AddPartitionResponse.cs [hunk 9]
```diff
@@ -3,16 +3,15 @@
...
 {
-    public AddPartitionResponse(PartitionEntry partitionEntry)
+    public AddPartitionResponse(Features.Partitioning.Partition partition)
     {
-        PartitionEntry = partitionEntry;
+        Partition = partition;
     }
 
-    public PartitionEntry PartitionEntry { get; }
+    public Features.Partitioning.Partition Partition { get; }
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Messages/Partitioning/AddPartitionResponse.cs:16]: Why is this necessary?

File: src/Microsoft.Health.Dicom.Core/Messages/Partitioning/GetPartitionResponse.cs [hunk 10]
```diff
@@ -3,16 +3,15 @@
...
 {
-    public GetPartitionResponse(PartitionEntry partitionEntry)
+    public GetPartitionResponse(Features.Partitioning.Partition partition)
     {
-        PartitionEntry = partitionEntry;
+        Partition = partition;
     }
 
-    public PartitionEntry PartitionEntry { get; }
+    public Features.Partitioning.Partition Partition { get; }
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Messages/Partitioning/GetPartitionResponse.cs:16]: What's this for?

File: swagger/v1-prerelease/swagger.yaml [hunk 11]
```diff
@@ -3219,18 +3219,15 @@ components:
...
       properties:
-        partitionKey:
+        key:
           type: integer
           format: int32
-        partitionName:
+        name:
           type: string
           nullable: true
-        createdDate:
```
Comment (wsugarman) [inline swagger/v1-prerelease/swagger.yaml:3231]: Yeah.... we'll need to add this back lol

Comment (wsugarman) [reply swagger/v1-prerelease/swagger.yaml:3231]: Oh dang, all of the names will need to be updated too, but I think we can isolate these changes to this code path

File: test/Microsoft.Health.Dicom.Web.Tests.E2E/Rest/ExtendedQueryTagTests.cs [hunk 12]
```diff
@@ -20,6 +20,7 @@
 using Microsoft.Health.Operations;
 using Xunit;
 using FunctionsStartup = Microsoft.Health.Dicom.Functions.App.Startup;
+using Partition = Microsoft.Health.Dicom.Core.Features.Partitioning.Partition;
```
Comment (wsugarman) [inline test/Microsoft.Health.Dicom.Web.Tests.E2E/Rest/ExtendedQueryTagTests.cs:23]: This is because of the client type collision?

**Review by wsugarman — CHANGES_REQUESTED** [review]

File: src/Microsoft.Health.Dicom.Api.UnitTests/Features/Routing/UrlResolverTests.cs [hunk 13]
```diff
@@ -123,7 +124,7 @@ public void GivenAnInstance_WhenRetrieveInstanceUriIsResolved_ThenCorrectUrlShou
 
         _httpContext.Request.RouteValues.Add(KnownActionParameterNames.Version, version);
 
-        var instance = new InstanceIdentifier(studyInstanceUid, seriesInstanceUid, sopInstanceUid);
+        var instance = new InstanceIdentifier(studyInstanceUid, seriesInstanceUid, sopInstanceUid, Partition.Default);
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Api.UnitTests/Features/Routing/UrlResolverTests.cs:127]: Oh that's right. Nevermind!

File: src/Microsoft.Health.Dicom.Client/Models/Partition.cs [hunk 14]
```diff
@@ -3,25 +3,33 @@
 // Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
 // -------------------------------------------------------------------------------------------------
 
+using System;
+using System.Text.Json.Serialization;
 using EnsureThat;
 
 namespace Microsoft.Health.Dicom.Client.Models;
 
+[Serializable]
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Client/Models/Partition.cs:12]: This shouldn't be necessary.

File: src/Microsoft.Health.Dicom.Client/Models/PartitionEntry.cs [hunk 15]
```diff
@@ -3,11 +3,16 @@
...
 
+using System;
+
 namespace Microsoft.Health.Dicom.Client.Models;
 
-public static class DefaultPartition
+/// <summary>
+/// Used to ensure API contract does not change while we change the underlying property naming.
+/// </summary>
+public class PartitionEntry : Partition
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Client/Models/PartitionEntry.cs:13]: Oh I left these comments in client. The function stuff is only relevant for Core lol

File: src/Microsoft.Health.Dicom.Core/Features/Routing/IUrlResolver.cs [hunk 16]
```diff
@@ -45,8 +45,9 @@ public interface IUrlResolver
     /// Resovles the URI to retrieve an instance.
     /// </summary>
     /// <param name="instanceIdentifier">The identifier to the instance.</param>
+    /// <param name="isPartitionEnabled">Whether or not partition feature is enabled</param>
     /// <returns>An instance of <see cref="Uri"/> pointing to where the instance can be retrieved.</returns>
-    Uri ResolveRetrieveInstanceUri(InstanceIdentifier instanceIdentifier);
+    Uri ResolveRetrieveInstanceUri(InstanceIdentifier instanceIdentifier, bool isPartitionEnabled);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Routing/IUrlResolver.cs:50]: Unsure how I feel about this. I like bubbling up the partition concept to the signature as opposed to the hidden HttpContext resolution. There are concepts like "studyInstanceUid` above that are UIDs without the context of a partition. Should there be an "identifier" class that is agnostic of partition? Or perhaps the better question is: how should the URI resolver consistently across its methods know whether to use a partition? Perhaps that is something determined by DI (e.g. a partitioned "UrlResolver" versus a non-partitioned?)

Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Features/Routing/IUrlResolver.cs:50]: Also would the method above `ResolveRetrieveStudyUri` need partition information?

File: src/Microsoft.Health.Dicom.Client/Models/Partition.cs [hunk 17]
```diff
@@ -9,17 +9,16 @@
...
     public const string DefaultName = "Microsoft.Default";
 
     public const int DefaultKey = 1;
 
-    [JsonPropertyName("PartitionKey")]
+    [JsonPropertyName("partitionKey")] // these explicit names are here for the REST schema
     public int Key { get; }
 
-    [JsonPropertyName("PartitionName")]
+    [JsonPropertyName("partitionName")] // these explicit names are here for the REST schema
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Client/Models/Partition.cs:21]: There must be a way for us to incorporate camelCase names as a setting for Swagger gen (like we can for JSON.NET and System.Text.Json), but I won't block on that here

File: src/Microsoft.Health.Dicom.Core/Features/Store/IStoreResponseBuilder.cs (hunk 7 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Features/Store/IStoreResponseBuilder.cs:32]: This still needs to move 😛

File: src/Microsoft.Health.Dicom.Client/Models/PartitionEntry.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Client/Models/PartitionEntry.cs:13]: At the very least, you must add `JsonProperty` attributes to the class for `Partition` as it's used in `ExportInput`. Functions JSON.NET and changing the name of the property would be a breaking change that would affect any in-progress functions.

File: src/Microsoft.Health.Dicom.Functions.UnitTests/Update/UpdateDurableFunctionTests.Activity.cs [hunk 18]
```diff
@@ -36,15 +36,15 @@ public async Task GivenInstanceMetadata_WhenUpdatingInstanceWatermark_ThenShould
 
         var versions = expected.Select(x => x.Version).ToList();
 
-        _indexStore.BeginUpdateInstancesAsync(Partition.DefaultKey, studyInstanceUid, CancellationToken.None).Returns(identifiers);
+        _indexStore.BeginUpdateInstancesAsync(Arg.Any<Partition>(), studyInstanceUid, CancellationToken.None).Returns(identifiers);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions.UnitTests/Update/UpdateDurableFunctionTests.Activity.cs:39]: nit: if this is the default partition, you should be able to pass it here as it will use reference equality on the type.

File: src/Microsoft.Health.Dicom.Core/Features/Partitioning/Partition.cs [hunk 19]
```diff
@@ -13,6 +13,8 @@ public class Partition
 {
     public const string DefaultName = "Microsoft.Default";
 
+    public const string UnknownName = "Unknown"; // use for older binaries where we are unable to get partition name and it is not known
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Partitioning/Partition.cs:16]: nit: maybe `Unspecified`?

File: src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV35.cs [hunk 20]
```diff
@@ -62,7 +63,7 @@ public override async Task<IReadOnlyList<InstanceMetadata>> RetrieveDeletedInsta
                                 seriesInstanceUid,
                                 sopInstanceUid,
                                 watermark,
-                                partitionKey),
+                                new Partition(partitionKey, Partition.UnknownName)),
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV35.cs:66]: nit: perhaps we can have a utility method for these legacy scenarios

**Comment by smithago** [comment]
For others looking at the review, can you help understand excatly want is the change that is resulting in enforcement? There are too many renames and it is unable to pin point the exact change

File: src/Microsoft.Health.Dicom.Client/Models/PartitionEntry.cs [hunk 21]
```diff
@@ -4,22 +4,15 @@
...
 using System;
-using EnsureThat;
 
 namespace Microsoft.Health.Dicom.Client.Models;
 
-public class PartitionEntry
+/// <summary>
+/// Used to ensure API contract does not change while we change the underlying property naming.
+/// </summary>
+public class PartitionEntry : Partition
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Client/Models/PartitionEntry.cs:13]: What's the point of this class now? I don't think anything operates on this type, nor does the type name matter here because we only cared about it for Swagger (even though that's not such a big deal either lol)

File: src/Microsoft.Health.Dicom.Api/Features/Routing/UrlResolver.cs [hunk 22]
```diff
@@ -109,7 +109,7 @@ public Uri ResolveRetrieveWorkitemUri(string workitemInstanceUid)
     }
 
     /// <inheritdoc />
-    public Uri ResolveRetrieveInstanceUri(InstanceIdentifier instanceIdentifier)
+    public Uri ResolveRetrieveInstanceUri(InstanceIdentifier instanceIdentifier, bool isPartitionEnabled)
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Api/Features/Routing/UrlResolver.cs:112]: Is this change required?

File: src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportSink.cs [hunk 23]
```diff
@@ -63,7 +63,7 @@ public async Task<bool> CopyAsync(ReadResult value, CancellationToken cancellati
             return false;
         }
 
-        using Stream sourceStream = await _source.GetStreamingFileAsync(value.Identifier.Version, value.Identifier.PartitionName, cancellationToken);
+        using Stream sourceStream = await _source.GetStreamingFileAsync(value.Identifier.Version, value.Identifier.Partition.Name, cancellationToken);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportSink.cs:66]: Are we sure that the Partition is populated in Export?

File: src/Microsoft.Health.Dicom.Core/Features/Partitioning/Partition.cs [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+using Newtonsoft.Json;
+
+namespace Microsoft.Health.Dicom.Core.Features.Partitioning;
+
+public class Partition
+{
+    public const string DefaultName = "Microsoft.Default";
+
+    public const string UnknownName = "Unknown"; // use for older binaries where we are unable to get partition name and it is not known
+
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Partitioning/Partition.cs:18]: Not sure why we want Unknown name. It should always be Microsoft.Default. There is no entry in DB that corresponds to Unknown Partition entry

File: src/Microsoft.Health.Dicom.Core/Features/Model/InstanceIdentifier.cs [hunk 25]
```diff
@@ -17,34 +17,12 @@ public class InstanceIdentifier
...
-        EnsureArg.IsNotNullOrWhiteSpace(seriesInstanceUid, nameof(seriesInstanceUid));
-        EnsureArg.IsNotNullOrWhiteSpace(sopInstanceUid, nameof(sopInstanceUid));
-
-        StudyInstanceUid = studyInstanceUid;
-        SeriesInstanceUid = seriesInstanceUid;
-        SopInstanceUid = sopInstanceUid;
-        PartitionKey = partitionKey;
-        PartitionName = partitionName;
-        PartitionEntry = new PartitionEntry(PartitionKey, PartitionName);
+        Partition = EnsureArg.IsNotNull(partition, nameof(partition));
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Model/InstanceIdentifier.cs:22]: Checking for null partition is correct, but are we ensuring that the key is not default?

File: src/Microsoft.Health.Dicom.Core/Features/Model/VersionedInstanceIdentifier.cs [hunk 26]
```diff
@@ -14,20 +14,8 @@ public class VersionedInstanceIdentifier : InstanceIdentifier
...
-
-    public VersionedInstanceIdentifier(
-        string studyInstanceUid,
-        string seriesInstanceUid,
-        string sopInstanceUid,
-        long version,
-        int partitionKey = DefaultPartition.Key,
-        string partitionName = DefaultPartition.Name)
-        : base(studyInstanceUid, seriesInstanceUid, sopInstanceUid, partitionKey, partitionName)
+        Partition partition = null)
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Model/VersionedInstanceIdentifier.cs:17]: In which scenario do you want to accept null partition?

File: src/Microsoft.Health.Dicom.Core/Features/Routing/IUrlResolver.cs (hunk 16 above)
Comment (bcarthic) [reply src/Microsoft.Health.Dicom.Core/Features/Routing/IUrlResolver.cs:50]: Why do we need this? Already the existing code populates the partition based on route values

File: src/Microsoft.Health.Dicom.Core/Features/Store/StoreResponseBuilder.cs [hunk 27]
```diff
@@ -97,12 +106,12 @@ private bool HasWarningReasonCode()
...
 
         var referencedSop = new DicomDataset
         {
             { DicomTag.ReferencedSOPInstanceUID, dicomDataset.GetSingleValue<string>(DicomTag.SOPInstanceUID) },
-            { DicomTag.RetrieveURL, _urlResolver.ResolveRetrieveInstanceUri(dicomInstance).ToString() },
+            { DicomTag.RetrieveURL, _urlResolver.ResolveRetrieveInstanceUri(dicomInstance, _isPartitionEnabled).ToString() },
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Store/StoreResponseBuilder.cs:114]: Do we need this? Even without this it populates the partition right?

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Activity.cs [hunk 28]
```diff
@@ -42,7 +43,7 @@ public async Task<IReadOnlyList<InstanceFileState>> UpdateInstanceWatermarkAsync
 
         logger.LogInformation("Beginning to update all instance watermarks");
 
-        IEnumerable<InstanceMetadata> instanceMetadata = await _indexStore.BeginUpdateInstancesAsync(arguments.PartitionKey, arguments.StudyInstanceUid, CancellationToken.None);
+        IEnumerable<InstanceMetadata> instanceMetadata = await _indexStore.BeginUpdateInstancesAsync(new Partition(arguments.PartitionKey, Partition.UnknownName), arguments.StudyInstanceUid, CancellationToken.None);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Activity.cs:47]: Is the unknown Partition Name correct? Even though the underlying method doesn't use the name, in future if its used will have issues. What is the problem in just accept the key in this case?

**Review by wsugarman — DISMISSED** [review]

File: src/Microsoft.Health.Dicom.SqlServer/Features/Partitioning/SqlPartitionStoreV6.cs [hunk 29]
```diff
@@ -85,18 +84,17 @@ public override async Task<IEnumerable<PartitionEntry>> GetPartitionsAsync(Cance
                        VLatest.Partition.PartitionName,
                        VLatest.Partition.CreatedDate);
 
-                    results.Add(new PartitionEntry(
+                    results.Add(new Partition(
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Partitioning/SqlPartitionStoreV6.cs:87]: If createdDate is not fetched, or you not changing the sproc?

File: src/Microsoft.Health.Dicom.Core/Features/Store/IIndexDataStore.cs [hunk 30]
```diff
@@ -139,20 +140,20 @@ public interface IIndexDataStore
     /// <summary>
     /// Asynchronously updates a DICOM instance NewWatermark
     /// </summary>
-    /// <param name="partitionKey">The partition key.</param>
+    /// <param name="partition">The partition.</param>
     /// <param name="versions">List of instances watermark to update</param>
     /// <param name="cancellationToken">The cancellation token.</param>
     /// <returns>A task that with list of instance metadata with new watermark.</returns>
-    Task<IEnumerable<InstanceMetadata>> BeginUpdateInstanceAsync(int partitionKey, IReadOnlyCollection<long> versions, CancellationToken cancellationToken = default);
+    Task<IEnumerable<InstanceMetadata>> BeginUpdateInstanceAsync(Partition partition, IReadOnlyCollection<long> versions, CancellationToken cancellationToken = default);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Store/IIndexDataStore.cs:147]: I am not sure why we needed this change?

File: src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV6.cs [hunk 31]
```diff
@@ -197,7 +198,7 @@ public override async Task<IEnumerable<VersionedInstanceIdentifier>> RetrieveDel
                             seriesInstanceUid,
                             sopInstanceUid,
                             watermark,
-                            partitionKey));
+                            new Partition(partitionKey, Partition.UnknownName)));
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV6.cs:201]: This is incorrect, since the corresponding partitionKey will not match Unknown Name. Either we need to pass the name which we have from the requestcontext or remove the old files

File: src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV35.cs [hunk 32]
```diff
@@ -62,7 +63,7 @@ public override async Task<IReadOnlyList<InstanceMetadata>> RetrieveDeletedInsta
...
                             new InstanceProperties()
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV35.cs:67]: Same here, Defaulting to uknown is wrong. Either populate the name or remove the files

File: src/Microsoft.Health.Dicom.Core/Features/Store/IIndexDataStore.cs (hunk 30 above)
Comment (bcarthic) [reply src/Microsoft.Health.Dicom.Core/Features/Store/IIndexDataStore.cs:147]: If that is the case, please pass the name why pass wrong value. You have the value in the FIlterAttribute, we just need to populate and pass on here

File: src/Microsoft.Health.Dicom.Core/Features/Model/InstanceIdentifier.cs (hunk 25 above)
Comment (bcarthic) [reply src/Microsoft.Health.Dicom.Core/Features/Model/InstanceIdentifier.cs:22]: Because there is some inconsistency in the model, you check for null for name but for key you dont check if it greater than 0

File: src/Microsoft.Health.Dicom.Core/Features/Model/VersionedInstanceIdentifier.cs (hunk 26 above)
Comment (bcarthic) [reply src/Microsoft.Health.Dicom.Core/Features/Model/VersionedInstanceIdentifier.cs:17]: Is there any repercussion of making this change? we are introducing a new change which wasn't there before

**Review by wsugarman — APPROVED** [review]
